File: apps/native/src/lib/watch-snapshot.ts
import type { ChatMessage, ChatProviderStatus } from "@VISP/api/chat/contract";
import type { StreamState, VideoConfiguration } from "../../modules/visp-srt";
import type { AudioTier } from "./audio-level";

export type WatchSnapshot = {
	version: 1;
	updatedAt: number;
	stream: {
		state: StreamState;
		liveStartedAt?: number;
		audioTier: AudioTier;
		width?: number;
		height?: number;
		fps?: number;
		message?: string;
		reconnectAttempt?: number;
	};
	chat: {
		statuses: Partial<Record<"twitch" | "kick", ChatProviderStatus["state"]>>;
		messages: Array<{
			id: string;
			provider: "twitch" | "kick";
			senderName: string;
			senderColor?: string;
			text: string;
		}>;
	};
	obs?: {
		configured: boolean;
		connected: boolean;
		pending: boolean;
		scenes: string[];
		currentScene: string | null;
		desiredScene: string | null;
	};
};

export function buildWatchSnapshot({
	audioTier,
	configuration,
	liveStartedAt,
	message,
	messages,
	obs,
	reconnectAttempt,
	state,
	statuses,
	updatedAt = Date.now(),
}: {
	audioTier: AudioTier;
	configuration?: VideoConfiguration;
	liveStartedAt?: number;
	message?: string;
	messages: ChatMessage[];
	obs?: WatchSnapshot["obs"];
	reconnectAttempt?: number;
	state: StreamState;
	statuses: Partial<Record<"twitch" | "kick", ChatProviderStatus["state"]>>;
	updatedAt?: number;
}): WatchSnapshot {
	return {
		version: 1,
		updatedAt,
		stream: {
			state,
			liveStartedAt,
			audioTier,
			width: configuration?.width,
			height: configuration?.height,
			fps: configuration?.fps,
			message,
			reconnectAttempt,
		},
		chat: {
			statuses,
			messages: messages.slice(-3).map((chatMessage) => ({
				id: chatMessage.id,
				provider: chatMessage.provider,
				senderName: chatMessage.sender.name,
				senderColor: chatMessage.sender.color,
				text: chatMessage.fragments.map((fragment) => fragment.text).join(""),
			})),
		},
		obs,
	};
}

File: apps/native/modules/visp-srt/ios/WatchSnapshot.swift
import Foundation

struct WatchSnapshot: Codable, Equatable {
  static let supportedVersion = 1

  let version: Int
  let updatedAt: Double
  let stream: WatchStreamSnapshot
  let chat: WatchChatSnapshot

  static func decode(_ data: Data) throws -> WatchSnapshot {
    let snapshot = try JSONDecoder().decode(WatchSnapshot.self, from: data)
    guard snapshot.version == supportedVersion else {
      throw WatchSnapshotError.unsupportedVersion(snapshot.version)
    }
    return snapshot
  }
}

enum WatchSnapshotError: Error, Equatable {
  case unsupportedVersion(Int)
}

struct WatchStreamSnapshot: Codable, Equatable {
  let state: String
  let liveStartedAt: Double?
  let audioTier: Int
  let width: Int?
  let height: Int?
  let fps: Int?
  let message: String?
  let reconnectAttempt: Int?
}

struct WatchChatSnapshot: Codable, Equatable {
  let statuses: [String: String]
  let messages: [WatchChatMessage]
}

struct WatchChatMessage: Codable, Equatable, Identifiable {
  let id: String
  let provider: String
  let senderName: String
  let senderColor: String?
  let text: String
}
